sphynx/networkit/additions.go

```diff
@@ -0,0 +1,55 @@
+// Additional code to make it easier to work with NetworKit.
+package networkit
+
+type SphynxId uint32
+type EdgeBundle struct {
+	Src         []SphynxId
+	Dst         []SphynxId
+	EdgeMapping []int64
+}
+
+type VertexSet struct {
+	MappingToUnordered []int64
+	MappingToOrdered   map[int64]SphynxId
+}
+
+func ToDoubleSlice(v DoubleVector) []float64 {
+	s := make([]float64, v.Size())
+	for i := range s {
+		s[i] = v.Get(i)
+	}
+	return s
+}
+
+func ToIdSlice(v IdVector) []SphynxId {
+	s := make([]SphynxId, v.Size())
+	for i := range s {
+		s[i] = SphynxId(v.Get(i))
+	}
+	return s
+}
+
+func ToNetworKit(vs VertexSet, es EdgeBundle) Graph {
+	builder := NewGraphBuilder(uint64(len(vs.MappingToUnordered)))
+	for i := range es.Src {
+		builder.AddHalfEdge(uint64(es.Src[i]), uint64(es.Dst[i]))
```

Comment:
> Half? What about the other half? :) That's somehow handled by `ToGraph` below I suppose?
> 
> And wow, this is indeed not no-copy. Wasn't we originally motivated by NetworKit data structures when designing Sphynx so that we don't need this?

Yeah I had the same shock. I looked a lot at NetworKit code, e.g. how their file readers create graphs. It's always extremely complex and uses unsafe APIs. ([NetworkitBinaryReader.cpp](https://github.com/networkit/networkit/blob/7.1/networkit/cpp/io/NetworkitBinaryReader.cpp)) Jano's PR used this GraphBuilder interface too and it's [documented](https://github.com/networkit/networkit/blob/7.1/include/networkit/graph/GraphBuilder.hpp) as the way to "speed up" building a graph.

> Or maybe that was only for attributes with this ordered/unordered thing?

Ah, I  think that's it! For attributes it's fairly straightforward.

> What is their internal represenation? Neighbor lists for both incoming and outgoing edges? Just guessing based on the half edge notation...

Yes.

```c++
std::vector<std::vector<node>> inEdges;
std::vector<std::vector<node>> outEdges;
```

> Maybe we should consider caching Networkit represenation of edge bundles as well?

Let's see how slow it is first. But caching would definitely make sense.